Sources/SwiftProtobuf/TextFormatEncoder.swift

```diff
@@ -33,57 +33,91 @@ private let tab = [UInt8](repeating: asciiSpace, count: tabSize)
 
 /// TextFormatEncoder has no public members.
 internal struct TextFormatEncoder {
+    private static let foyerSize = 4096
+    private var foyer = Array<UInt8>(repeating: 0, count: TextFormatEncoder.foyerSize)
+    private var foyerIndex = 0
+
     private var data = [UInt8]()
     private var indentString: [UInt8] = []
-    var stringResult: String {
-        get {
-            return String(bytes: data, encoding: String.Encoding.utf8)!
+
+    internal mutating func constructFinalResult() -> String {
+        if foyerIndex > 0 {
+            let subfoyer = foyer[0..<foyerIndex]
+            data.append(contentsOf: subfoyer)
+            foyerIndex = 0
+        }
+        return String(bytes: data, encoding: String.Encoding.utf8)!
+    }
+
+    // Append each byte to the foyer, append and
+    // recycle the foyer when it fills up.
+    private mutating func append(byte: UInt8) {
+        if foyerIndex > TextFormatEncoder.foyerSize - 1 {
+            data.append(contentsOf: foyer)
+            foyerIndex = 0
         }
+        foyer[foyerIndex] = byte
+        foyerIndex += 1
     }
 
     internal mutating func append(staticText: StaticString) {
         let buff = UnsafeBufferPointer(start: staticText.utf8Start, count: staticText.utf8CodeUnitCount)
-        data.append(contentsOf: buff)
+        for b in buff {
```

Comment:
These loops (this and the next few functions) look like potential bottlenecks again, should they flush and directly add to `data` or maybe use some helper that copies chunks that fit into the buffer instead?

Reply:
For the particular example found by OSS-Fuzz, almost all of the appends are one or just a couple of bytes.  It appears that in this case, the overhead of appending to data outweighs the cost of doing repeated single-byte stores into the array.

Reply:
I'd expect that repeated calls to `append(_:)` would generally still be slower than a single `append(contentsOf:)` call, _especially_ if the source buffer is just an UnsafeBufferPointer.

`append(_:)` isn't (supposed to be) magically cheaper than chunking...